core/src/main/java/org/frankframework/parameters/Parameter.java

```diff
@@ -832,16 +836,22 @@ private Object getValueForFormatting(ParameterValueList alreadyResolvedParameter
 		Object substitutionValue = paramValue == null ? null : paramValue.getValue();
 
 		if (substitutionValue == null) {
-			Message substitutionValueMessage = session.getMessage(name);
-			if (!substitutionValueMessage.isEmpty()) {
-				if (substitutionValueMessage.asObject() instanceof Date) {
-					substitutionValue = preFormatDateType(substitutionValueMessage.asObject(), formatType, formatString);
+			Object substitutionValueMessage = session.get(name);
+			if (substitutionValueMessage != null) {
+				if (substitutionValueMessage instanceof Date) {
+					substitutionValue = preFormatDateType(substitutionValueMessage, formatType, formatString);
 				} else {
-					try {
-						substitutionValue = substitutionValueMessage.asString();
-					} catch (IOException e) {
-						throw new ParameterException(getName(), "Cannot get substitution value", e);
+					if (substitutionValueMessage instanceof String) {
+						substitutionValue = substitutionValueMessage;
+					} else {
+						try {
+							Message message = Message.asMessage(substitutionValueMessage); // Do not close Message from session here; might be reused later
+							substitutionValue = message.asString();
+						} catch (IOException e) {
+							throw new ParameterException(getName(), "Cannot get substitution value from session key: " + name, e);
+						}
```

Comment:
I think we should close it here in a try-with-resources, but that we should add an extra else-if to check if `substitionValueMessage instanceof Message`, which case to we not close it.
That seems safer to me wrt. avoiding leaks.